### PR TITLE
Adding support for Zstd Compression

### PR DESCRIPTION
**Pull request by IAL32**
The current API of `TFRecordOptions` only supports:
- no compression
- GZIP compression
- ZLIB compression

With this PR, I have added support for a 4th compression algorithm, ZSTD: https://github.com/facebook/zstd/

The API for `TFRecordOptions` is therefore updated to accept the following parameters:

```
      compression_type: `"GZIP"`, `"ZLIB"`, `"ZSTD"` or `""` (no compression).
      zstd_input_buffer_size: int or `None`. 
      zstd_output_buffer_size: int or `None`.
      zstd_flush_mode: int or `None`.
      zstd_nb_workers: int or `None`.
      zstd_compression_level: 0 to 19, or `None`. Default: ZSTD_CLEVEL_DEFAULT.
      zstd_compression_strategy: 0 to 9, or `None`. Default: 0.
      zstd_window_log: between ZSTD_WINDOWLOG_MIN and ZSTD_WINDOWLOG_MAX. Default: 0.
```

Maintaining backward compatibility with the current API.

*Note: the branch is called `adrian-compresssion-r2.6`, as originally it was based on r2.6, but it has been since updated to match current master branch 3038529326066c4eb5cb0453d7f82d929551a24e*

### CODE REVIEW

**Comment by mihaimaruseac** [comment]
Please open feature PRs against master.

**Comment by IAL32** [comment]
> Please open feature PRs against master.

I am still actively working on the PR, and until I have a working version, I am sticking to a stable release as origin.
Nevertheless, thanks for the heads up!

**Comment by mihaimaruseac** [comment]
Given the amount of changes you will have a lot of conflicts when you migrate to master. Note that we never merge release branches back into master.

**Comment by mihaimaruseac** [comment]
Also, CI on the release branch is not the same as on master, so you should not rely on its results

**Comment by IAL32** [comment]
@mihaimaruseac I have finished and refined the PR contents, and it is now ready for review :)

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by IAL32** [comment]
I am curious, why does the `Code Check - Changed Files` step fail? When clicking into the details it shows "The specified bucket is missing".
Also `import/copybara` fails, and there are no details as to why.
Is there anything I am missing and should I take action to correct these failures?

Thanks a lot in the meantime!

**Comment by mihaimaruseac** [comment]
Oh, the CI setup changed on master recently. Can you rebase to the top of the branch, please?

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by IAL32** [comment]
@mihaimaruseac I think I have a mismatch of `clang-format` version, as executing `clang-format --style=Google -i tensorflow/python/lib/io/record_io_wrapper.cc` outputs a different result from the one in the Code Check step.

I am using Ubuntu 18.04 with `clang-format --version`:
```
clang-format version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
``` 

Any way of knowing which clang-format version it's used in the CI step?

**Comment by mihaimaruseac** [comment]
C++ linter is complaining:

```
No message!
(in test file /usertools/code_check_changed_files.bats, line 36)
  `[[ ! -s $BATS_TEST_TMPDIR/needs_help.txt ]]' failed
clang-format is recommended. Here are the suggested changes:
=============================
diff --git a/tensorflow/python/lib/io/record_io_wrapper.cc b/-
index e9d403da44d..00000000000 100644
--- a/tensorflow/python/lib/io/record_io_wrapper.cc
+++ b/-
@@ -356,16 +356,17 @@ PYBIND11_MODULE(_pywrap_record_io, m) {
            [](PyRecordWriter* self, py::args) {
              MaybeRaiseRegisteredFromStatus(self->Close());
            })
-      .def("write",
-           [](PyRecordWriter* self, tensorflow::StringPiece record) {
-             tensorflow::Status status;
-             {
-               py::gil_scoped_release release;
-               status = self->WriteRecord(record);
-             }
-             MaybeRaiseRegisteredFromStatus(status);
-           },
-           py::arg("record"))
+      .def(
+          "write",
+          [](PyRecordWriter* self, tensorflow::StringPiece record) {
+            tensorflow::Status status;
+            {
+              py::gil_scoped_release release;
+              status = self->WriteRecord(record);
+            }
+            MaybeRaiseRegisteredFromStatus(status);
+          },
+          py::arg("record"))
       .def("flush",
            [](PyRecordWriter* self) {
              MaybeRaiseRegisteredFromStatus(self->Flush());
You can use clang-format --style=Google -i <file> to apply changes to a file.
```

File: tensorflow/python/lib/io/record_io_wrapper.cc [hunk 1]
```diff
@@ -316,11 +316,26 @@ PYBIND11_MODULE(_pywrap_record_io, m) {
...
+                     &ZstdCompressionOptions::compression_strategy)
+      .def_readwrite("window_log", &ZstdCompressionOptions::window_log);
+
   using tensorflow::io::RecordWriterOptions;
   py::class_<RecordWriterOptions>(m, "RecordWriterOptions")
       .def(py::init(&RecordWriterOptions::CreateRecordWriterOptions))
       .def_readonly("compression_type", &RecordWriterOptions::compression_type)
-      .def_readonly("zlib_options", &RecordWriterOptions::zlib_options);
+      .def_readonly("zlib_options", &RecordWriterOptions::zlib_options)
+      .def_readonly("zstd_options", &RecordWriterOptions::zstd_options);
```
[the remark this answers is not in the document]
Comment (IAL32) [reply tensorflow/python/lib/io/record_io_wrapper.cc:338]: It isn't, one has `zlib_options` and the other `zstd_options`

[the remark this answers is not in the document]
Comment (mihaimaruseac) [reply tensorflow/python/lib/io/record_io_wrapper.cc:338]: My bad, I failed to properly read it.

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @IAL32 Can you please check build failures. Thank you!

**Comment by IAL32** [comment]
Hmm, I have the following errors I don't currently understand:
- [PyLint](https://github.com/tensorflow/tensorflow/runs/5818563713?check_suite_focus=true) - `The config file tensorflow/tools/ci_build/pylintrc doesn't exist!`.

I don't think this has anything to do with my changes, I'll rebase and see what happens.

- [Community CI Build](http://ml-ci.amd.com:21096/job/tensorflow/job/github-prs-upstream-master/job/AMD-ROCm-Community-CI-Build/job/PR-53385/29/display/redirect) - `ERROR: The project you're trying to build requires Bazel 5.1.0 (specified in /workspace/.bazelversion), but it wasn't found in /usr/local/lib/bazel/bin.`

This seems to be a problem with the build agent not having the required bazel binary.

About the other failures, I am working on them, but it may take a while.

Thanks!

**Review by mihaimaruseac — APPROVED** [review]

**Comment by IAL32** [comment]
https://github.com/IAL32/TUMWS21GR

In case anyone is interested, this is part of a guided research project @ TUM. I have written a report detailing preprocessing pipeline speeds using this implementation of Zstandard.

**Comment by gbaned** [comment]
Hi @IAL32 Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @IAL32 Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
Hi @IAL32 Any update on this PR? Please. Thank you!

**Pull request closed without merging**